### PR TITLE
[WIP] fix(es/transforms/typescript): enum and namespace declaration block scoping

### PR DESCRIPTION
**Pull request by dsherret**
I've started on #2886 this also solves #2801

### CODE REVIEW

File: crates/swc/benches/typescript.rs [hunk 1]
```diff
@@ -48,16 +48,18 @@ fn parse(c: &swc::Compiler) -> (Arc<SourceFile>, Program) {
...
 }
 
-fn as_es(c: &swc::Compiler) -> Program {
-    let program = parse(c).1;
-
-    program.fold_with(&mut typescript::strip())
+fn as_es(program: Program) -> Program {
+    let mark = Mark::fresh(Mark::root());
+    program
+        .fold_with(&mut ts_resolver(mark))
```
[the remark this answers is not in the document]
Comment (kdy1) [reply crates/swc/benches/typescript.rs:58]: If so, we should make `resolver` aware of declaration merging.
Maybe it's already aware of it, (considering the operation is performs), but it requires double-check.

File: crates/swc/src/config/mod.rs [hunk 2]
```diff
@@ -356,6 +358,8 @@ impl Options {
             );
 
         let pass = chain!(
+            Optional::new(resolver_with_mark(top_level_mark), !syntax.typescript()),
+            Optional::new(ts_resolver(top_level_mark), syntax.typescript()),
```
[the remark this answers is not in the document]
Comment (kdy1) [reply crates/swc/src/config/mod.rs:362]: `ts_resolver` is sort of optimization for non-typescript code, but yeah, maybe `resolver` is enough.

[the remark this answers is not in the document]
Comment (kdy1) [reply crates/swc/src/config/mod.rs:362]: I made https://github.com/swc-project/swc/pull/2931 which will remove `ts_resolver`.

[the remark this answers is not in the document]
Comment (kdy1) [reply crates/swc/src/config/mod.rs:362]: We should change pass order if we are going to use `ts_resolver`.
Currently `strip` is invoked before `resolver`.

**Comment by dsherret** [comment]
I'm going to redo this PR since I've started splitting things up into separate PRs. I will submit an update soon.

**Pull request closed without merging**